### PR TITLE
[Backend] Add registerExecutor API on backend register

### PR DESCRIPTION
**Pull request by struss**
This commit will add `registerExcutor` API on `backendRegistrationApi`.

`registerExecutor` API will use to register executor and use this on Device API.

ONE-vscode-DCO-1.0-Signed-off-by: struss <rrstrous@nate.com>

### CODE REVIEW

File: src/Backend/Backend.ts [hunk 1]
```diff
@@ -29,10 +30,12 @@ interface BackendMap {
 
 // List of backend extensions registered
 let globalBackendMap: BackendMap = {};
+// List of Executor extensions registered
+let globalExecutorArray: Executor[] = [];
 
 function backendRegistrationApi() {
   const logTag = 'backendRegistrationApi';
-  let registrationAPI = {
+  return {
```
Comment (struss) [inline src/Backend/Backend.ts:38]: This change occur by `SonarLint`, typescript:S1488 rule.
```
Local variables should not be declared and then immediately returned or thrown (typescript:S1488)
```

Comment (struss) [reply src/Backend/Backend.ts:38]: I think this change will not related with this PR. I will rebase this.

**Review by YongseopKim — APPROVED** [review]
LGTM

File: src/Tests/Backend/Backend.test.ts [hunk 2]
```diff
@@ -35,7 +35,14 @@ class BackendMockup implements Backend {
...
+
+const executorName = 'Mockup';
+class ExecutorMockup extends ExecutorBase {
+  name(): string {
+    return executorName;
+  }
+}
 
 suite('Backend', function() {
   suite('backendRegistrationApi', function() {
```
Comment (yunjayh) [inline src/Tests/Backend/Backend.test.ts:48]: Check after backend registration? (Did I understood the code correctly?)

```ts
assert.strictEqual(globalExecutorArray.length, 1);
```

File: src/Tests/Backend/Backend.test.ts [hunk 3]
```diff
@@ -64,5 +84,6 @@ suite('Backend', function() {
     if (gToolchainEnvMap[backendName] !== undefined) {
       delete gToolchainEnvMap[backendName];
     }
+    globalExecutorArray.length = 0;
```
[the remark this answers is not in the document]
Comment (struss) [reply src/Tests/Backend/Backend.test.ts:87]: Um... I will update this.

File: src/Tests/Backend/Backend.test.ts (hunk 2 above)
Comment (struss) [reply src/Tests/Backend/Backend.test.ts:48]: Oh I will add this. Thanks!

**Review by YongseopKim — APPROVED** [review]

File: src/Backend/Backend.ts [hunk 4]
```diff
@@ -41,11 +44,19 @@ function backendRegistrationApi() {
       if (compiler) {
         gToolchainEnvMap[backend.name()] = new ToolchainEnv(compiler);
       }
+      const executor = backend.executor();
+      if (executor) {
+        globalExecutorArray.push(executor);
+      }
       Logger.info(logTag, 'Backend', backendName, 'was registered into ONE-vscode.');
+    },
+    registerExecutor(executor: Executor) {
```
Comment (dayo09) [inline src/Backend/Backend.ts:53]: Q. Why is executor registered aside from Backend..? I thought it would included in backend.

Comment (struss) [reply src/Backend/Backend.ts:53]: I thought there could be a Executor which is not related to Backend, so I make a room for this.

Comment (struss) [reply src/Backend/Backend.ts:53]: but, if `Backend` could get multiple `Executor`, I think we don't keep this, as every `Executor` have each related `Backend`.

Comment (struss) [reply src/Backend/Backend.ts:53]: I think #1076 change will remove this api as multiple executor could registerd on.

**Pull request merged**